trace-utils/src/msgpack_decoder/v05/mod.rs

```diff
@@ -0,0 +1,401 @@
+// Copyright 2024-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::msgpack_decoder::decode::error::DecodeError;
+use crate::msgpack_decoder::decode::{
+    map::read_map_len,
+    number::read_number_bytes,
+    string::{handle_null_marker, read_string_bytes},
+};
+use crate::span::SpanBytes;
+use std::collections::HashMap;
+
+const PAYLOAD_LEN: u32 = 2;
+const SPAN_ELEM_COUNT: u32 = 12;
+
+/// Decodes a slice of bytes into a vector of `TracerPayloadV05` objects.
+///
+///
+///
+/// # Arguments
+///
+/// * `data` - A tinybytes Bytes buffer containing the encoded data. Bytes are expected to be
+///   encoded msgpack data containing a list of a list of v05 spans.
+///
+/// # Returns
+///
+/// * `Ok(Vec<TracerPayloadV05>)` - A vector of decoded `TracerPayloadV05` objects if successful.
+/// * `Err(DecodeError)` - An error if the decoding process fails.
+///
+/// # Errors
+///
+/// This function will return an error if:
+/// - The array length for trace count or span count cannot be read.
+/// - Any span cannot be decoded.
+///
+/// # Examples
+///
+/// ```
+/// use datadog_trace_utils::msgpack_decoder::v05::from_slice;
+/// use rmp_serde::to_vec;
+/// use std::collections::HashMap;
+/// use tinybytes;
+///
+/// let data = (
+///     vec!["".to_string()],
+///     vec![vec![(
+///         0,
+///         0,
+///         0,
+///         1,
+///         2,
+///         3,
+///         4,
+///         5,
+///         6,
+///         HashMap::<u32, u32>::new(),
+///         HashMap::<u32, f64>::new(),
+///         0,
+///     )]],
+/// );
+/// let encoded_data = to_vec(&data).unwrap();
+/// let encoded_data_as_tinybytes = tinybytes::Bytes::from(encoded_data);
+/// let (decoded_traces, _payload_size) =
+///     from_slice(encoded_data_as_tinybytes).expect("Decoding failed");
+///
+/// assert_eq!(1, decoded_traces.len());
+/// assert_eq!(1, decoded_traces[0].len());
+/// let decoded_span = &decoded_traces[0][0];
+/// assert_eq!("", decoded_span.name.as_str());
+/// ```
+pub fn from_slice(mut data: tinybytes::Bytes) -> Result<(Vec<Vec<SpanBytes>>, usize), DecodeError> {
+    let data_elem = rmp::decode::read_array_len(unsafe { data.as_mut_slice() })
+        .map_err(|_| DecodeError::InvalidFormat("Unable to read payload len".to_string()))?;
+
+    if data_elem != PAYLOAD_LEN {
+        return Err(DecodeError::InvalidFormat(
+            "Invalid payload size".to_string(),
+        ));
+    }
+
+    let dict = deserialize_dict(&mut data)?;
+
+    let trace_count = rmp::decode::read_array_len(unsafe { data.as_mut_slice() })
+        .map_err(|_| DecodeError::InvalidFormat("Unable to read trace len".to_string()))?;
```

Comment:
I realize you are by and large copying behavior from v04, so I don't think this is necessarily blocking the PR. But after this PR is merged we should probably audit all the places where we are mapping errors and not including the original error message and change the behavior to provide more detail in our error messages.